src/term/check/has_entrypoint.rs

```diff
@@ -1,32 +1,89 @@
 use crate::{
-  term::{Book, Name},
+  term::{Book, Definition, Name},
   ENTRY_POINT, HVM1_ENTRY_POINT,
 };
+use std::fmt::Display;
+
+#[derive(Debug, Clone)]
+pub enum EntryErr {
+  NotFound(Name),
+  Multiple(Vec<Name>),
+  MultipleRules,
+  Arguments,
+}
+
+impl Display for EntryErr {
+  fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+    match self {
+      EntryErr::NotFound(name) => write!(f, "File has no '{name}' definition"),
+      EntryErr::Multiple(fnd) if fnd.len() == 2 => {
+        write!(f, "File has both '{}' and '{}' definitions", fnd[0], fnd[1])
+      }
+      EntryErr::Multiple(fnd) => {
+        write!(f, "File has '{}', '{}' and '{}' definitions", fnd[0], fnd[1], fnd[2])
+      }
+      EntryErr::MultipleRules => write!(f, "Main definition can't have more than one rule"),
+      EntryErr::Arguments => write!(f, "Main definition can't have any arguments"),
+    }
+  }
+}
 
 impl Book {
-  pub fn check_has_entrypoint(&mut self, nam: Option<Name>) -> Result<Name, String> {
-    if let Some(nam) = nam {
-      match self.defs.get(&nam) {
-        Some(_) => {
-          self.entrypoint = Some(nam.clone());
-          return Ok(nam);
+  pub fn check_has_entrypoint(&mut self) -> Option<Name> {
```

Comment:
Adding to what Kate said, I think this would be better as a transformation pass `set_entrypoint`, where we set `Book.entrypoint` to either `main`, `Main` or some custom entrypoint depending on the options and the function definitions (or None in case of error i guess).

Reply:
What about now? or would it still be better to not set the `book.entrypoint` field on `main.rs` at all and just pass around as an arg until the `set_entrypoint` is called?

Reply:
They way you did it is good